src/tests/ftest/container/query_attribute.py

```diff
@@ -0,0 +1,113 @@
+        query. From the output, verify the pool/container UUID matches the one
+        that was returned when creating the pool/container.
+
+        Attribute test:
+        1. Create 5 attributes to a container. We can set attribute
+        name and corresponding value, so set unique name-value pair for each of
+        the 5 attributes.
+        2. Call continaer list-attrs. This returns the 5 attribute names, so
+        compare them against the actual names used.
+        3. Call container get-attr. This returns the value of the corresponding
+        name, so do this for each of the 5 attributes.
+
+    :avocado: recursive
+    """
+
+    def test_container_query_attr(self):
+        """JIRA ID: DAOS-4640
+
+        Test Description:
+            Test daos container query and attribute commands as described
+            above.
+
+        Use Cases:
+            Test container query, set-attr, list-attr, and get-attr commands.
+
+        :avocado: tags=all,pool,small,full_regression,cont_query_attr
+        """
+        # 1. Test pool query.
+        kwargs = {"scm_size": "1G"}
+        pool_create_result = self.get_dmg_command().get_output(
+            "pool_create", **kwargs)
+        expected_pool_uuid = pool_create_result[0]
+        sr = pool_create_result[1]
+        daos_cmd = DaosCommand(self.bin)
+        # Create container and store the UUID as expected.
+        kwargs = {"pool": expected_pool_uuid, "svc": sr}
+        expected_cont_uuid = daos_cmd.get_output(
+            "container_create", **kwargs)[0]
+        # Call daos container query, obtain pool and container UUID, and
+        # compare against those used when creating the pool and the container.
+        kwargs["cont"] = expected_cont_uuid
+        query_output = daos_cmd.get_output("container_query", **kwargs)[0]
+        actual_pool_uuid = query_output[0]
+        actual_cont_uuid = query_output[1]
+        self.assertEqual(actual_pool_uuid, expected_pool_uuid)
+        self.assertEqual(actual_cont_uuid, expected_cont_uuid)
+
+        # 2. Test container set-attr, get-attr, and list-attrs.
+        expected_attrs = []
+        expected_attrs_dict = {}
+        sample_attrs = []
+        sample_vals = []
+        # Create 5 attributes.
+        for i in range(5):
+            sample_attr = "attr" + str(i)
+            sample_val = "val" + str(i)
+            sample_attrs.append(sample_attr)
+            sample_vals.append(sample_val)
+            _ = daos_cmd.container_set_attr(
+                pool=actual_pool_uuid, cont=actual_cont_uuid,
+                attr=sample_attr, val=sample_val, svc=sr).stdout
+            expected_attrs.append(sample_attr)
+            expected_attrs_dict[sample_attr] = sample_val
+        expected_attrs.sort()
+        # List the attribute names.
+        kwargs = {
+            "pool": actual_pool_uuid,
+            "svc": sr,
+            "cont": actual_cont_uuid
```

Comment:
Since we have a single key with two possibly different values - expected and actual - would it make sense to have a dictionary of dictionaries:
```
# Initialize the list of container attributes with expected values
attributes = {
    "attr{}".format(key): {
        "expected": "val{}".format(key), "actual": None} for key in range(5)}

# Set/get the actual values for the container attributes
errors = []
for key in attributes:
    daos_cmd.container_set_attr(
        pool=actual_pool_uuid, cont=actual_cont_uuid,
        attr=key, val=attribute[key]["expected"], svc=sr)
    attribute[key]["actual"] = daos_cmd.get_output(
        "container_get_attr", pool=actual_pool_uuid,
        cont=actual_cont_uuid, attr=key, svc=sr)[0]
    if attribute[key]["expected"] != attribute[key]["actual"]:
        errors.append(
            "Unexpected output for daos container get-attr {}: {} != {}".format(
                key, attribute[key]["actual"], attribute[key]["expected"]))
self.assertEqual(0, len(errors), "; ".join(errors))

# Verify all the attributes are returned in the list
attribute_name_list = daos_cmd.get_output(
    "container_list_attrs", pool=actual_pool_uuid, svc=sr, cont=actual_cont_uuid)
self.assertEqual(
    list(attributes.keys()).sort(), attribute_name_list.sort(),
    "Unexpected output from daos container list-attrs")
```

Reply:
Also do we need to verify different/invalid attribute names and values?  Should the names/values come from the test yaml file?

Reply:
Should we test some limit to the number of attributes that can be set?

Reply:
Thanks for the suggestion. I thought more about possible test cases and added the following cases.

Alphabet only, number only, alphabet and number, long string, special characters, and string that includes space.

I also added a test case to set 50 attributes.

If the names/values come from the yaml file, the server will be restarted for each test case and it takes longer, so I think it would be better to test everything in one test case.

For invalid names and values, it looks like daos doesn’t have any restriction. If I pass in some special character like “<”, the error is generated by the console and not daos. E.g.,

[mkano@wolf-3 ~]$ daos container set-attr --pool=094872f9-5eb0-4f36-9cd0-14f4b47efbba --svc=0 --cont=0a7920ca-3dfb-499d-a7fa-a22da25a8b5f --attr=attr1 --val=abc<
-bash: syntax error near unexpected token `newline'

So it’s probably not relevant and it’s a little awkward to test because sometimes the console waits for signal depending on the special character we use.